lib/events.js

```diff
@@ -404,9 +405,17 @@ EventEmitter.prototype.listeners = function listeners(type) {
     return [];
 
   if (typeof evlistener === 'function')
-    return [evlistener.listener || evlistener];
+    return unwrap ? [evlistener.listener || evlistener] : [evlistener];
+
+  return unwrap ? unwrapListeners(evlistener) : arrayClone(evlistener);
+}
+
+EventEmitter.prototype.listeners = function listeners(type) {
+  return _listeners(this, type, true);
+};
 
-  return unwrapListeners(evlistener);
+EventEmitter.prototype.rawListeners = function rawListeners(type) {
```

Comment:
Is there a way to not make this public? I don't like to expose internal details of the `EventEmitter`.

Reply:
I guess the question is whether it's beneficial to hide these internals or not. I've already seen lots of user-code reaching into `_events` for the wrappers. We do it ourselves even. That seems to indicate that we probably need a way to get those `once` wrappers.

Reply:
FWIW there's good discussion on this in https://github.com/nodejs/node/pull/6881 where the change `listeners` was made. It seems like @addaleax & @Fishrock123 wanted to add an ability to return the wrapped listeners anyway but then that fell by the wayside.

Reply:
> I've already seen lots of user-code reaching into `_events` for the wrappers

Can you link some examples which is not `readable-stream`? We added `eventNames()` to prevent people from using `_events` and there weren't a lot of modules using `_events` at the time.

> We do it ourselves even. That seems to indicate that we probably need a way to get those once wrappers.

Agreed but I would prefer to have it private.

Reply:
I've seen it in private codebases for this particular use case, described by @ChALkeR in the PR linked above:

> It looks like there will be no way to re-attach the events then, which could be useful for cloning, intercepting, or temporary disabling events. Perhaps some API should be added that allows that?

As an example out in the wild, this module broke after the change to `listeners` and would need to use `_events` now. (Of course, as far as I can tell, `prependListener` just solves the same problem anyway but that's a different story.)

https://github.com/timoxley/overshadow-listeners/blob/master/index.js

Also, isn't Ultron currently subtly broken:

https://github.com/unshiftio/ultron/blob/336c789616db9ca3f164f4e5f15ed78f49d528da/index.js#L99-L111

Reply:
It seems a weak definition of **lots** of user-code :)

That hack in `ultron` was needed when once events weren't unwrapped. It's still there to not break support for older Node.js versions.

That said I'm not blocking or anything, I just find it odd to expose that method to everyone.

Reply:
I mean, I've seen it in lots of literal user code not npm modules. Usually it's just the need to temporarily detach events and then re-attach. I've seen code that detaches all listeners to wrap them in a wrapper to do some logging & diagnostics.

I'm not super passionate about having this be publicly available or anything. If the consensus is it's not useful, we can make it internal only.

Reply:
Yeah, I think this should be exposed for API completeness (for the reason in the @ChALkeR quote above); and that there is code in Node core that uses this should be a good indicator that there are valid use cases.

Reply:
I don't see it coming, but changes to how once events are handled can potentially translate to breaking changes if this api is exposed.

Reply:
@lpinca can you please articulate this more? What would change on how `once` events are handled?
I think we will have breaking changes anyway if people used `_events`, just outside of the public API.

Reply:
This will probably never happen but imagine that we will switch to an object to handle events, for example:

```js
{ once: false, listener }
```

`rawListeners` can no longer return a function.

> I think we will have breaking changes anyway if people used _events, just outside of the public API.

Yes, the difference is that `_events` is "private" so you know beforehand that if you use it your code can break at any time.

Reply:
However we can't currently touch `_events` because a good chunk of the ecosystem uses it.